### ISSUE TITLE
Stokhos: Amesos2 solvers are not working for GPU backend with more than one MPI rank

### ISSUE THREAD

**Issue by romintomasetti** [trilinos/Trilinos#12279]
## Bug Report

@etphipp

### Description

`Amesos2` solvers (at least `KLU2` and `Basker`) do work for ensemble types with several ranks **only if there is no device**.

That is, for ensemble types with `Cuda` or `HIP` **and** with more than one rank, trying to solve fails. On a single rank, solving with ensemble type works.

The problem is due to `Tpetra::Details::WrappedDualView` counting (the infamous `Cannot access data on host while a device view is alive` problem).

For instance, the problem pops up in `KLU2` at the end of `solve_impl` when doing the `do_put` (see below provided `gdb` backtrace). However, I think it pops up there but the problem must be looked for in some upstream code. Indeed, for non-ensemble type, we only have the `Amesos2` stuff. For ensemble type, `Stokhos` uses the `create_flat_vector_view` (and related) to "flatten" the system but if I get it correctly, `Amesos2` is then called without any further manipulation of the data.

I think (but not sure) that when running with 2 ranks, the problem only happens on the non-root rank. That might be an hint.
Also, the fact that the solvers work in the same conditions for non-ensemble types is another hint.

If you want to have a reproducer, I suggest to have a look at @maartenarnst 's reproducer in #11578 (https://github.com/trilinos/Trilinos/issues/11578#issuecomment-1476705395), and simply change from a `double` scalar type to `Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, STOKHOS_DEFAULT_ENSEMBLE_SIZE, execution_space>>`. You should see the test fails with `WrappedDualView` throwing at you :wink:

### Steps to Reproduce
1. SHA1: 608f54c97a294ddbbaeb69f992402a62ddfb654e
1. Shake fist angrily at computer (done :heavy_check_mark: )

### GDB backtrace

Here is the relevant part of my `gdb` output when running a test with a command like `mpiexec -np 2 gdb -ex run -ex bt <executable-path>`. This test works for `double` and `std::complex<double>` but fails as mentioned above for ensemble types.

```bash
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[ RUN      ] TestFixtureWithExecutionSpace_execution_space_Timoshenko1DTest.TestAgainstExactSol
terminate called after throwing an instance of 'std::runtime_error'
  what():  Tpetra::Details::WrappedDualView (name = MV::DualView; host use_count = 2; device use_count = 4): Cannot access data on host while a device view is alive

Thread 1 "exec_Timoshenko" received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140734698422272) at ./nptl/pthread_kill.c:44
44      ./nptl/pthread_kill.c: No such file or directory.
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140734698422272) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140734698422272) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140734698422272, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007fffd72b8476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007fffd729e7f3 in __GI_abort () at ./stdlib/abort.c:79
--Type <RET> for more, q to quit, c to continue without paging--#5  0x00007fffd7627b9e in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007fffd763320c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x00007fffd7633277 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#8  0x00007fffd76334d8 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00007fffddf14ca1 in Tpetra::Details::WrappedDualView<Kokkos::DualView<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, void> >::throwIfDeviceViewAlive (this=0x5555578f3880) at /home/costmo-user/Trilinos/packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:684
#10 0x00007fffddeff54f in Tpetra::Details::WrappedDualView<Kokkos::DualView<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, void> >::getHostView (this=0x5555578f3880) at /home/costmo-user/Trilinos/packages/tpetra/core/src/Tpetra_Details_WrappedDualView.hpp:263
#11 0x00007fffddeddd5f in Tpetra::MultiVector<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >::getLocalViewHost (
    this=0x5555578f3720, s=...) at /home/costmo-user/Trilinos/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp:3980
#12 0x00007fffddef070a in Tpetra::MultiVector<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >::unpackAndCombine (
    this=0x5555578f3720, importLIDs=..., imports=..., constantNumPackets=1, CM=Tpetra::REPLACE, space=...)
    at /home/costmo-user/Trilinos/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp:2057
#13 0x00007fffdf5eb21e in Tpetra::DistObject<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >::doUnpackAndCombine
    (this=0x5555578f3720, remoteLIDs=..., constantNumPackets=1, CM=Tpetra::REPLACE, space=...)
    at /home/costmo-user/Trilinos/packages/tpetra/core/src/Tpetra_DistObject_def.hpp:1664
#14 0x00007fffdf5e8a31 in Tpetra::DistObject<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >::endTransfer (
    this=0x5555578f3720, src=..., transfer=..., modeString=0x7fffffffb5b0 "endImport (forward mode)", 
    revOp=Tpetra::DistObject<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >::DoForward, CM=Tpetra::REPLACE, 
    restrictedMode=false) at /home/costmo-user/Trilinos/packages/tpetra/core/src/Tpetra_DistObject_def.hpp:1339
#15 0x00007fffdf5e199d in Tpetra::DistObject<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >::endImport (
    this=0x5555578f3720, source=..., importer=..., CM=Tpetra::REPLACE, restrictedMode=false)
 at /home/costmo-user/Trilinos/packages/tpetra/core/src/Tpetra_DistObject_def.h--Type <RET> for more, q to quit, c to continue without paging--   pp:549
#16 0x00007fffdf5dff34 in Tpetra::DistObject<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >::doImport (
    this=0x5555578f3720, source=..., importer=..., CM=Tpetra::REPLACE, restrictedMode=false)
    at /home/costmo-user/Trilinos/packages/tpetra/core/src/Tpetra_DistObject_def.hpp:298
#17 0x00007fffbb3e2c35 in Amesos2::MultiVecAdapter<Tpetra::MultiVector<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> > >::put1dData_kokkos_view<Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::OpenMP> > (this=0x55555b6c55d0, kokkos_new_data=..., lda=0, source_map=..., 
    distribution=Amesos2::ROOTED) at /home/costmo-user/Trilinos/packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp:588
#18 0x00007fffbb3c9699 in Amesos2::Util::put_1d_data_helper_kokkos_view<Amesos2::MultiVecAdapter<Tpetra::MultiVector<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> > >, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::OpenMP> >::do_put (mv=..., kokkos_data=..., ldx=0, 
    distribution_map=..., distribution=Amesos2::ROOTED) at /home/costmo-user/Trilinos/packages/amesos2/src/Amesos2_MultiVecAdapter_def.hpp:373
#19 0x00007fffbb3a6e3b in Amesos2::Util::put_1d_data_helper_kokkos_view<Amesos2::MultiVecAdapter<Tpetra::MultiVector<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> > >, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::OpenMP> >::do_put (mv=..., kokkos_data=..., ldx=0, 
    distribution=Amesos2::ROOTED, indexBase=0) at /home/costmo-user/Trilinos/packages/amesos2/src/Amesos2_MultiVecAdapter_def.hpp:394
#20 0x00007fffbb37532e in Amesos2::KLU2<Tpetra::CrsMatrix<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >, Tpetra::MultiVector<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> > >::solve_impl (this=0x55555b6dcc90, X=..., B=...)
    at /home/costmo-user/Trilinos/packages/amesos2/src/Amesos2_KLU2_def.hpp:383
#21 0x00007fffbb525bf1 in Amesos2::SolverCore<Amesos2::KLU2, Tpetra::CrsMatrix<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >, Tpetra::MultiVector<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> > >::solve (this=0x55555b6dcc90, 
    X=..., B=...) at /home/costmo-user/Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp:226
#22 0x00007fffbb5259dc in Amesos2::SolverCore<Amesos2::KLU2, Tpetra::CrsMatrix<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> >, Tpetra::MultiVector<double, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaSpace> > >::solve (this=0x55555b6dcc90)
    at /home/costmo-user/Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp:175
```

@maartenarnst

**Comment by maartenarnst** [comment]
Closing as solved in pr #12324.

**Comment by maartenarnst** [comment]
Hi @romintomasetti, would you have a second to close it?
